### PR TITLE
[feature] #1790: Improve performance by using stack-based vectors.

### LINKED ISSUE #1790: Use stack-based vectors.
We often use a vector as a substitute for a built-in list for very few elements (e.g. one), and since the data is being used externally: serialized and deserialized, neither heap elision, nor register compaction can take place. `TinyVec` is a crate that allows dynamic placement with small-array optimisation: putting small vectors on the stack, while handling larger collections on the heap.

### PR DESCRIPTION
**Pull request by appetrosyan**
### Description of the Change

Incorporate the indirect dependencies: `smallvec` and `smallstr` into main iroha code. 
Measured a performance uplift of 10% ± 1%, measured using `torii` benchmark. 

Added extra crate for data primitives not directly related to `iroha_data_model`, called `iroha_data_primitives`. 

### Issue

Closes #1790.

### Benefits
More performance

### Possible Drawbacks
 More work for library developers. 
 
 ### Tests
 
 ```
 cd client
 cargo bench # cargo bench torii will result in no runs executed. 
 ```
 
 #### Before 
 
image

#### After

image

### CODE REVIEW

**Comment by appetrosyan** [comment]
The code seems to trigger a `cargo nono` false-positive, on the `smallvec` and `smallstr` crates. The program can be compiled with `no_std` support, and `smallvec` is already a third-degree dependency of Iroha. As a temporary workaround, I will just move the `iroha_data_model::small` module into a different workspace, with a comment to move it back as soon as `cargo nono` fixes [#61](https://github.com/hobofan/cargo-nono/issues/61).

File: data_model/src/transaction.rs [hunk 1]
```diff
@@ -61,16 +62,34 @@ pub trait Txn {
...
     /// WebAssembly smartcontract
     Wasm(Vec<u8>),
 }
 
+/// Wrapper for byte representation of [`Executable::Wasm`].  Inline
+/// into [`Executable::Wasm`] as soon as GATs are stabilised and
+/// implementations for from byte vector can be split off from
+/// [`IntoIterator<Item = Instruction>`].
+pub struct WasmSmartContract {
+    raw_data: small::SmallVec<[u8; 32]>,
```
[the remark this answers is not in the document]
Comment (appetrosyan) [reply data_model/src/transaction.rs:78]: The point of this is that `32` bytes is the size that we get "for free", due to the average size of the other variant. If we need significantly more space for wasm blobs, then I'll need to revert this to a plain heap-allocated `Vec`, as that would catastrophically impact memory usage.

[the remark this answers is not in the document]
Comment (mversic) [reply data_model/src/transaction.rs:78]: When compiled from rust, Wasm blob contains 600 - 700 kB, I don't expect it to be any better for other language sources.

Maybe some handwritten compiled wasms would fit into 32 byes. This is hard for me to tell, 32 bytes seems too small

[the remark this answers is not in the document]
Comment (mversic) [reply data_model/src/transaction.rs:78]: It can be a `Box<[u8]>` because the given blob will never be modified/resized

[the remark this answers is not in the document]
Comment (appetrosyan) [reply data_model/src/transaction.rs:78]: In that case I think it's better to revert it back to `vec<u8>`, or a heap-based array.

File: actor/src/broker.rs [hunk 2]
```diff
@@ -131,7 +131,7 @@ impl Broker {
                 }
             })
             .collect::<FuturesUnordered<_>>()
-            .collect::<Vec<_>>()
+            .collect::<smallvec::SmallVec<[_; 8]>>()
```
Comment (mversic) [inline actor/src/broker.rs:134]: is this related to `usize`?

File: client/Cargo.toml [hunk 3]
```diff
@@ -31,6 +31,9 @@ rand = "0.8"
 serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
 tungstenite = { version = "0.16" }
+smallvec = { version = "1.6.1", features = ["serde", "union"] }
+smallstr = { version = "0.2.0", features = ["serde", "union"] }
```
Comment (mversic) [inline client/Cargo.toml:35]: why use these in client? is there a need to optimize the client?

File: client/tests/integration_tests/add_asset.rs [hunk 4]
```diff
@@ -115,7 +115,9 @@ fn client_add_asset_with_decimal_should_increase_asset_amount() -> Result<()> {
         )),
     );
     // and check that it is added without errors
-    let sum = quantity.checked_add(quantity2)?;
+    let sum = quantity
+        .checked_add(quantity2)
+        .map_err(|e| eyre::eyre!("{}", e))?;
```
Comment (mversic) [inline client/tests/integration_tests/add_asset.rs:120]: Doesn't this just wraps it into `eyre::Report` twice?

File: crypto/src/lib.rs [hunk 5]
```diff
@@ -3,6 +3,7 @@
 #![cfg_attr(not(feature = "std"), no_std)]
 
 #[cfg(not(feature = "std"))]
+#[macro_use]
```
Comment (mversic) [inline crypto/src/lib.rs:6]: was this because `vec!` was missing? and maybe `format!` as well

File: version/src/lib.rs [hunk 6]
```diff
@@ -120,7 +120,7 @@ pub struct UnsupportedVersion {
     /// Version of the content.
     pub version: u8,
     /// Raw content.
-    pub raw: RawVersioned,
+    pub raw: Box<RawVersioned>,
```
Comment (mversic) [inline version/src/lib.rs:123]: what is the reason this is boxed now?

File: client/Cargo.toml (hunk 3 above)
Comment (appetrosyan) [reply client/Cargo.toml:35]: It does help the tests and both are indirect imports, so why not.

File: data_model/src/small.rs [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+    }
+}
+
+impl<A: Array> core::ops::DerefMut for SmallVec<A> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<A: Array> Eq for SmallVec<A> where A::Item: PartialEq + Eq {}
```
Comment (mversic) [inline data_model/src/small.rs:147]:
```suggestion
impl<A: Array> Eq for SmallVec<A> where A::Item: Eq {}
```

File: data_model/src/small.rs [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    A::Item: PartialOrd + Ord,
+{
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
+
+impl<A: Array> Ord for SmallVec<A>
+where
+    A::Item: PartialOrd + Ord,
```
Comment (mversic) [inline data_model/src/small.rs:126]:
```suggestion
    A::Item: Ord,
```

File: data_model/src/small.rs [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    A::Item: PartialEq + Eq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl<A: Array> PartialOrd for SmallVec<A>
+where
+    A::Item: PartialOrd + Ord,
```
Comment (mversic) [inline data_model/src/small.rs:117]:
```suggestion
    A::Item: Ord,
```

File: data_model/src/small.rs [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+    where
+        D: Deserializer<'de>,
+    {
+        deserialize_from_smallvec(deserializer)
+    }
+}
+
+impl<A: Array> PartialEq for SmallVec<A>
+where
+    A::Item: PartialEq + Eq,
```
Comment (mversic) [inline data_model/src/small.rs:108]:
```suggestion
    A::Item: Eq,
```

File: data_model/src/small.rs [hunk 11]
```diff
@@ -0,0 +1,231 @@
+//! Small-string optimisation-related implementations and
```
Comment (mversic) [inline data_model/src/small.rs:1]: this is a duplicate?

File: actor/src/broker.rs (hunk 2 above)
Comment (appetrosyan) [reply actor/src/broker.rs:134]: Actually, CPU cores. Sadly there can be no way of getting that information at compile time, so 8 seemed like a good guess.

File: data_model/src/lib.rs [hunk 12]
```diff
@@ -355,6 +357,19 @@ impl Value {
     }
 }
 
+impl<A: small::Array> From<SmallVec<A>> for Value
+where
+    A::Item: Into<Value>,
+{
```
Comment (mversic) [inline data_model/src/lib.rs:363]: do we still need `From<Vec<V>> for Value`?
and what about `TryFrom<Value> for SmallVec<A>`?

File: data_model/primitives/src/small.rs [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+/// ```
+pub struct SmallVec<A: Array>(pub smallvec::SmallVec<A>);
+
+impl<A: Array> Default for SmallVec<A> {
+    fn default() -> Self {
+        Self(smallvec::SmallVec::new())
+    }
+}
+
+impl<A: Array> Clone for SmallVec<A>
```
Comment (mversic) [inline data_model/primitives/src/small.rs:77]: you can derive it

File: data_model/primitives/src/small.rs [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+///
+/// To instantiate a vector with `N` stack elements,
+/// ```ignore
+/// use iroha_data_model::small::SmallVec;
+///
+/// let a: SmallVec<[u8; 24]> = SmallVec(smallvec::smallvec![32]);
+/// ```
+pub struct SmallVec<A: Array>(pub smallvec::SmallVec<A>);
+
+impl<A: Array> Default for SmallVec<A> {
```
Comment (mversic) [inline data_model/primitives/src/small.rs:71]: you can derive it

File: data_model/primitives/src/small.rs [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+impl<A: Array> Clone for SmallVec<A>
+where
+    A::Item: Clone,
+{
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl<A: Array> fmt::Debug for SmallVec<A>
```
Comment (mversic) [inline data_model/primitives/src/small.rs:86]: why not derive it?

File: data_model/primitives/src/small.rs [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+/// elements on the stack if `self.len()` is less than `N`, if not,
+/// produces a heap-allocated vector.
+///
+/// To instantiate a vector with `N` stack elements,
+/// ```ignore
+/// use iroha_data_model::small::SmallVec;
+///
+/// let a: SmallVec<[u8; 24]> = SmallVec(smallvec::smallvec![32]);
+/// ```
+pub struct SmallVec<A: Array>(pub smallvec::SmallVec<A>);
```
Comment (mversic) [inline data_model/primitives/src/small.rs:69]: why didn't you derive `PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize`?

**Review by mversic — CHANGES_REQUESTED** [review]

File: client/Cargo.toml (hunk 3 above)
Comment (mversic) [reply client/Cargo.toml:35]: I wouldn't like to bother the user with this. As long as it's not evident in the API I'm fine with this.

File: crypto/src/lib.rs (hunk 5 above)
Comment (appetrosyan) [reply crypto/src/lib.rs:6]: Yeah.

File: actor/src/broker.rs (hunk 2 above)
Comment (mversic) [reply actor/src/broker.rs:134]: how is # of CPU cores related to stack size of the vec?
would it make sense to extract into a type or a const?

File: version/src/lib.rs (hunk 6 above)
Comment (appetrosyan) [reply version/src/lib.rs:123]: Size of enum. Without it, the variant is huge compared to the other variant.

Comment (mversic) [reply version/src/lib.rs:123]: as far as I can see that's not true:
```
pub enum RawVersioned {
    /// In JSON format.
    Json(String),
    /// In Parity Scale Codec format.
    ScaleBytes(Vec<u8>),
}
```

File: data_model/primitives/src/small.rs (hunk 14 above)
Comment (appetrosyan) [reply data_model/primitives/src/small.rs:71]: See my comment about `Clone`.

File: version/src/lib.rs (hunk 6 above)
Comment (mversic) [reply version/src/lib.rs:123]: it's just two extra bytes?

File: data_model/primitives/src/small.rs (hunk 16 above)
Comment (appetrosyan) [reply data_model/primitives/src/small.rs:69]: The standard: for the same reason I didn't derive `Copy` and `Debug`, because a `SmallVec` of non-`Ord` objects is reasonable, just not `Ord`. 

I know that it's possible to derive `serde` traits for structs with non-(de)serializable fields, but that is only one line shorter than what I did, also requiring that the `SmallVec` only hold types satisfying a certain bound. Same as above, I want `SmallVec` to be (de)serializable only if what I store in it also is, and have the option to store things that are not (de)serializable.

File: data_model/src/lib.rs (hunk 12 above)
Comment (appetrosyan) [reply data_model/src/lib.rs:363]: > do we still need From<Vec<V>> for Value?

Yes. We can't replace `Vec` everywhere, because `SmallVec<Value>` would be recursive, if `Value::Vec` wasn't heap-allocated, and `Box` defeats the purpose. 

> TryFrom<Value> for SmallVec<A>

Good point. Will add shortly.

File: core/src/queue.rs [hunk 17]
```diff
@@ -434,6 +434,7 @@ mod tests {
         let max_block_tx = 2;
         let alice_key = KeyPair::generate().expect("Failed to generate keypair.");
         let wsv = WorldStateView::new(world_with_test_domains(alice_key.public_key.clone()));
+        println!("{:?}", wsv);
```
[the remark this answers is not in the document]
Comment (appetrosyan) [reply core/src/queue.rs:445]: It's a test, but you're right.

File: data_model/primitives/src/small.rs (hunk 13 above)
Comment (mversic) [reply data_model/primitives/src/small.rs:77]: ah, yes, I get it. I missed a

File: client/tests/integration_tests/add_asset.rs (hunk 4 above)
Comment (appetrosyan) [reply client/tests/integration_tests/add_asset.rs:120]: Nope. Turning `FixedPointOperationError`, (which doesn't `impl std::error::Error`, because of `#[no_std]`), into `eyre::Report`.

File: data_model/primitives/src/small.rs (hunk 16 above)
Comment (mversic) [reply data_model/primitives/src/small.rs:69]: yes, sorry, I was aware of this problem, I just didn't see it here

File: data_model/primitives/src/small.rs [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+    type IntoIter = <smallvec::SmallVec<A> as IntoIterator>::IntoIter;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<T: IntoSchema, A: smallvec::Array<Item = T>> IntoSchema for SmallVec<A> {
+    fn type_name() -> String {
+        "Vec<".to_owned() + &T::type_name() + ">"
```
Comment (mversic) [inline data_model/primitives/src/small.rs:205]:
```suggestion
        Vec::<T>::type_name()
```

File: data_model/primitives/src/small.rs [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+    }
+
+    fn schema(map: &mut MetaMap) {
+        let _ = map
+            .entry(Self::type_name())
+            .or_insert_with(|| Metadata::Vec(T::type_name()));
+        if !map.contains_key(&T::type_name()) {
+            T::schema(map);
```
Comment (mversic) [inline data_model/primitives/src/small.rs:213]:
```suggestion
    Vec::<T>::schema(map);
```

File: data_model/primitives/src/small.rs [hunk 20]
```diff
@@ -0,0 +1,251 @@
...
+        }
+    }
+}
+
+impl IntoSchema for SmallStr {
+    fn type_name() -> String {
+        String::from("String")
+    }
+    fn schema(map: &mut MetaMap) {
+        let _ = map.entry(Self::type_name()).or_insert(Metadata::String);
```
Comment (mversic) [inline data_model/primitives/src/small.rs:223]:
```suggestion
        String::schema(map);
```

File: data_model/primitives/src/small.rs [hunk 21]
```diff
@@ -0,0 +1,251 @@
...
+            .or_insert_with(|| Metadata::Vec(T::type_name()));
+        if !map.contains_key(&T::type_name()) {
+            T::schema(map);
+        }
+    }
+}
+
+impl IntoSchema for SmallStr {
+    fn type_name() -> String {
+        String::from("String")
```
Comment (mversic) [inline data_model/primitives/src/small.rs:220]:
```suggestion
        String::type_name()
```

**Review by mversic — CHANGES_REQUESTED** [review]

File: version/src/lib.rs [hunk 22]
```diff
@@ -9,7 +9,7 @@
 extern crate alloc;
 
 #[cfg(not(feature = "std"))]
-use alloc::{string::String, vec::Vec};
+use alloc::{boxed::Box, string::String, vec::Vec};
```
Comment (mversic) [inline version/src/lib.rs:12]: ci is not complaining?

File: data_model/primitives/src/small.rs [hunk 23]
```diff
@@ -0,0 +1,259 @@
...
+        where
+            D: Deserializer<'de>,
+        {
+            deserialize_from_smallvec(deserializer)
+        }
+    }
+
+    impl<A: Array> PartialEq for SmallVec<A>
+    where
+        A::Item: PartialEq + Eq,
```
Comment (mversic) [inline data_model/primitives/src/small.rs:142]:
```suggestion
        A::Item: Eq,
```

should be implied?

File: data_model/primitives/src/small.rs [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+    {
+        fn eq(&self, other: &Self) -> bool {
+            self.0 == other.0
+        }
+    }
+
+    impl<A: Array> PartialOrd for SmallVec<A>
+    where
+        A::Item: PartialOrd + Ord,
```
Comment (mversic) [inline data_model/primitives/src/small.rs:151]:
```suggestion
        A::Item: Ord,
```
should be implied?

File: data_model/primitives/src/small.rs [hunk 25]
```diff
@@ -0,0 +1,259 @@
...
+    {
+        fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+            self.0.partial_cmp(&other.0)
+        }
+    }
+
+    impl<A: Array> Ord for SmallVec<A>
+    where
+        A::Item: PartialOrd + Ord,
```
Comment (mversic) [inline data_model/primitives/src/small.rs:160]:
```suggestion
        A::Item: Ord,
```
bound is implied

File: data_model/src/transaction.rs [hunk 26]
```diff
@@ -79,16 +79,34 @@ pub trait Txn {
...
 /// Either ISI or Wasm binary
-#[derive(
-    Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, FromVariant, IntoSchema,
-)]
+#[derive(Debug, Clone, PartialEq, Eq, Decode, Encode, Deserialize, Serialize, IntoSchema)]
 pub enum Executable {
     /// Ordered set of instructions.
     Instructions(Vec<Instruction>),
     /// WebAssembly smartcontract
     Wasm(Vec<u8>),
```
Comment (mversic) [inline data_model/src/transaction.rs:87]: shouldn't this be
```suggestion
    Wasm(WasmSmartContract),
```

Comment (mversic) [reply data_model/src/transaction.rs:87]: otherwise I don't see `WasmSmartContract` used anywhere else

File: data_model/primitives/src/fixed.rs [hunk 27]
```diff
@@ -0,0 +1,251 @@
...
+mod tests {
+    #![allow(clippy::restriction, clippy::panic)]
+    use super::*;
+
+    #[test]
+    fn cannot_parse_negative_value() {
+        match Fixed::try_from(-123.45_f64) {
+            Err(FixedPointOperationError::NegativeValue(_)) => (),
+            _ => panic!("Negative values shouldn't be allowed"),
+        }
```
Comment (s8sato) [inline data_model/primitives/src/fixed.rs:174]: Does this not work for?
```rust
assert!(matches!(Fixed::try_from(-123.45_f64), Err(FixedPointOperationError::NegativeValue(_))));
```

File: data_model/primitives/src/small.rs (hunk 23 above)
Comment (appetrosyan) [reply data_model/primitives/src/small.rs:142]: It should be `PartialEq`. Not `Eq` here.

File: data_model/primitives/src/small.rs (hunk 24 above)
Comment (appetrosyan) [reply data_model/primitives/src/small.rs:151]: Should be `PartialOrd` instead of `Ord`.

File: data_model/src/transaction.rs (hunk 26 above)
Comment (appetrosyan) [reply data_model/src/transaction.rs:87]: Rebase artifact.

File: core/src/genesis.rs [hunk 28]
```diff
@@ -23,8 +24,8 @@ use crate::{
     IrohaNetwork,
 };
 
-type Online = Vec<PeerId>;
-type Offline = Vec<PeerId>;
+type Online = SmallVec<[PeerId; 8]>;
+type Offline = SmallVec<[PeerId; 8]>;
```
[the remark this answers is not in the document]
Comment (appetrosyan) [reply core/src/genesis.rs:28]: 8 is optimal for tests. I've added a `TODO:` into the code, and opened #1855. The point is that we need to refine the estimates based off of real-world usage statistics. If we know the average case size of an object and memory budget, increasing this value might help significantly.

**Review by mversic — APPROVED** [review]

**Review by s8sato — APPROVED** [review]
